pkg/joinserver/grpc.go

```diff
@@ -30,7 +31,7 @@ func (srv jsServer) GetJoinEUIPrefixes(ctx context.Context, _ *pbtypes.Empty) (*
 	prefixes := make([]ttnpb.JoinEUIPrefix, 0, len(srv.JS.euiPrefixes))
 	for _, p := range srv.JS.euiPrefixes {
 		prefixes = append(prefixes, ttnpb.JoinEUIPrefix{
-			JoinEUI: &p.EUI64,
+			JoinEUI: p.EUI64.Copy(&types.EUI64{}),
```

Comment:
Why?

Reply:
Otherwise the reference to the actual stored `EUI64` passed and that may get mutated (and it is when the RPC is called).

If you want to see it for yourself - apply this diff to the PR:
```
diff --git a/pkg/joinserver/grpc.go b/pkg/joinserver/grpc.go
index 6c793df05..5ff6d3a6a 100644
--- a/pkg/joinserver/grpc.go
+++ b/pkg/joinserver/grpc.go
@@ -19,7 +19,6 @@ import (
 
        pbtypes "github.com/gogo/protobuf/types"
        "go.thethings.network/lorawan-stack/pkg/ttnpb"
-       "go.thethings.network/lorawan-stack/pkg/types"
 )
 
 type jsServer struct {
@@ -31,7 +30,7 @@ func (srv jsServer) GetJoinEUIPrefixes(ctx context.Context, _ *pbtypes.Empty) (*
        prefixes := make([]ttnpb.JoinEUIPrefix, 0, len(srv.JS.euiPrefixes))
        for _, p := range srv.JS.euiPrefixes {
                prefixes = append(prefixes, ttnpb.JoinEUIPrefix{
-                       JoinEUI: p.EUI64.Copy(&types.EUI64{}),
+                       JoinEUI: &p.EUI64,
                        Length:  uint32(p.Length),
                })
        }
```

And do `go test` in `pkg/joinserver`. Observe:
```
FAIL | --- FAIL: TestGetJoinEUIPrefixes (0.01s)
FAIL |     --- FAIL: TestGetJoinEUIPrefixes/Defined_JoinEUIPrefixes_Set_1 (0.00s)
     |         doc.go:66: 
     |             /home/rvolosatovs/src/go.thethings.network/lorawan-stack/pkg/joinserver/grpc_test.go:94
     |             Diff:
     |                Prefixes[0].JoinEUI[0]: 255 != 66
     |                Prefixes[0].JoinEUI[1]: 66 != 255
     |                Prefixes[0].JoinEUI[7]: 255 != 0
     |                Prefixes[1].JoinEUI[0]: 16 != 66
     |                Prefixes[1].JoinEUI[1]: 0 != 255
     |                Prefixes[1].JoinEUI[7]: 255 != 0
     |             Diff:     '(*ttnpb.JoinEUIPrefixes){Prefixes:[]ttnpb.JoinEUIPrefix{ttnpb.JoinEUIPrefix{JoinEUI:(*[8]uint8){255, 66, 255, 255, 255, 255, 255, 255, 0}, Length:12, XXX_NoUnkeyedLiteral:struct {}{}, XXX_sizecache:0}, ttnpb.JoinEUIPrefix{JoinEUI:(*[8]uint8){166, 0, 255, 255, 255, 255, 255, 255, 0}, Length:12, XXX_NoUnkeyedLiteral:struct {}{}, XXX_sizecache:0}, ttnpb.JoinEUIPrefix{JoinEUI:(*[8]uint8){66, 255, 255, 255, 255, 255, 255, 0}, Length:56, XXX_NoUnkeyedLiteral:struct {}{}, XXX_sizecache:0}}, XXX_NoUnkeyedLiteral:struct {}{}, XXX_sizecache:0}'
FAIL |     --- FAIL: TestGetJoinEUIPrefixes/Defined_JoinEUIPrefixes_Set_2 (0.01s)
     |         doc.go:66: 
     |             /home/rvolosatovs/src/go.thethings.network/lorawan-stack/pkg/joinserver/grpc_test.go:94
     |             Diff:
     |                Prefixes[0].JoinEUI[0]: 175 != 17
     |                Prefixes[0].JoinEUI[1]: 178 != 255
     |                Prefixes[0].JoinEUI[3]: 0 != 255
     |                Prefixes[0].JoinEUI[4]: 79 != 17
     |                Prefixes[0].JoinEUI[5]: 153 != 255
     |                Prefixes[0].JoinEUI[6]: 117 != 17
     |                Prefixes[0].JoinEUI[7]: 1 != 0
     |                Prefixes[1].JoinEUI[0]: 16 != 17
     |                Prefixes[1].JoinEUI[1]: 0 != 255
     |                Prefixes[1].JoinEUI[2]: 0 != 17
     |                Prefixes[1].JoinEUI[3]: 0 != 255
     |                Prefixes[1].JoinEUI[4]: 0 != 17
     |                Prefixes[1].JoinEUI[5]: 0 != 255
     |                Prefixes[1].JoinEUI[6]: 0 != 17
     |             Diff:     '(*ttnpb.JoinEUIPrefixes){Prefixes:[]ttnpb.JoinEUIPrefix{ttnpb.JoinEUIPrefix{JoinEUI:(*[8]uint8){175, 178255, 17, 0255, 179, 12535, 117, 10}, Length:1, XXX_NoUnkeyedLiteral:struct {}{}, XXX_sizecache:0}, ttnpb.JoinEUIPrefix{JoinEUI:(*[8]uint8){167, 0255, 017, 0255, 017, 0255, 017, 0}, Length:40, XXX_NoUnkeyedLiteral:struct {}{}, XXX_sizecache:0}, ttnpb.JoinEUIPrefix{JoinEUI:(*[8]uint8){17, 255, 17, 255, 17, 255, 17, 0}, Length:56, XXX_NoUnkeyedLiteral:struct {}{}, XXX_sizecache:0}}, XXX_NoUnkeyedLiteral:struct {}{}, XXX_sizecache:0}'
FAIL |     --- FAIL: TestGetJoinEUIPrefixes/Defined_JoinEUIPrefixes_Set_3 (0.00s)
     |         doc.go:66: 
     |             /home/rvolosatovs/src/go.thethings.network/lorawan-stack/pkg/joinserver/grpc_test.go:94
     |             Diff:
     |                Prefixes[0].JoinEUI[0]: 18 != 69
     |                Prefixes[0].JoinEUI[1]: 52 != 255
     |                Prefixes[0].JoinEUI[2]: 86 != 255
     |                Prefixes[0].JoinEUI[3]: 120 != 255
     |                Prefixes[0].JoinEUI[4]: 144 != 0
     |                Prefixes[0].JoinEUI[5]: 18 != 0
     |                Prefixes[0].JoinEUI[6]: 52 != 0
     |                Prefixes[0].JoinEUI[7]: 86 != 0
     |                Prefixes[1].JoinEUI[0]: 0 != 69
     |                Prefixes[1].JoinEUI[1]: 0 != 255
     |                Prefixes[1].JoinEUI[2]: 0 != 255
     |                Prefixes[1].JoinEUI[3]: 0 != 255
     |                Prefixes[1].JoinEUI[4]: 255 != 0
     |                Prefixes[1].JoinEUI[5]: 255 != 0
     |                Prefixes[1].JoinEUI[6]: 255 != 0
     |                Prefixes[1].JoinEUI[7]: 69 != 0
     |             Diff:     '(*ttnpb.JoinEUIPrefixes){Prefixes:[]ttnpb.JoinEUIPrefix{ttnpb.JoinEUIPrefix{JoinEUI:(*[8]uint8){1869, 5255, 86255, 12055, 1440, 180, 520, 860}, Length:4, XXX_NoUnkeyedLiteral:struct {}{}, XXX_sizecache:0}, ttnpb.JoinEUIPrefix{JoinEUI:(*[8]uint8){0, 0, 0, 069, 255, 255, 255, 690, 0, 0, 0}, Length:8, XXX_NoUnkeyedLiteral:struct {}{}, XXX_sizecache:0}, ttnpb.JoinEUIPrefix{JoinEUI:(*[8]uint8){69, 255, 255, 255, 0, 0, 0, 0}, Length:16, XXX_NoUnkeyedLiteral:struct {}{}, XXX_sizecache:0}}, XXX_NoUnkeyedLiteral:struct {}{}, XXX_sizecache:0}'
```

Reply:
We can also give `join_eui` in `JoinEUIPrefix` a `nullable=false` option.

Reply:
Indeed, but I believe we would still need to copy here.
I'll try this approach.

Reply:
Ah, I missed that it was a pointer. That is the core problem imo. Indeed, I don't think it should be a pointer and we should pass this by value

Reply:
That should fix it: https://github.com/TheThingsNetwork/lorawan-stack/pull/1248/commits/5a6a4c53bea37ccf571d147a99b92b4c1409cc68